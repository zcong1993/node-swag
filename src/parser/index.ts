export * from './definitions'
export * from './comments'
