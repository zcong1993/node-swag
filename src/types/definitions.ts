export interface Definition {
  type?: string
  required?: string[]
  items?: Definition
  enum?: string[]
  $ref?: string
  properties?: {
    [key: string]: Definition
  }
}

export type Definitions = {
  [key: string]: Definition
}
