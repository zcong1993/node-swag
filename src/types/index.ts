export * from './definitions'
export * from './rootConfig'
