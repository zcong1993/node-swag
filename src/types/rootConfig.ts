export interface BasicAuth {
  type: string
}

export interface ApiKeyAuth {
  type: string
  in: string
  name: string
}

export type Security = BasicAuth | ApiKeyAuth

export const isBasicAuth = (t: Security): t is BasicAuth => {
  return 'type' in t && !('in' in t)
}

export const isApiKeyAuth = (t: Security): t is ApiKeyAuth => {
  return 'name' in t && 'in' in t
}

interface Info {
  title?: string
  version?: string
  description?: string
  termsOfService?: string
  license?: {
    name?: string
    url?: string
  }
  contract?: {
    email?: string
    name?: string
    url?: string
  }
}

export interface rootConfig {
  host?: string
  basePath?: string
  info?: Info
  securityDefinitions?: {
    [key: string]: Security
  }
}
