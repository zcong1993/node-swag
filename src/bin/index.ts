#!/usr/bin/env node
import { join } from 'path'
import { existsSync, writeFileSync } from 'fs'
import * as cac from 'cac'
import { sync as mkdirp } from 'mkdirp'
import { stringify } from 'json2yaml'
import { rootConfig } from '../types'
import { gen } from '../'

const normalizePath = (p: string, root: string = process.cwd()): string =>
  join(root, p)

const cli = cac('swag')

cli
  .option('--config <config>', 'Config file', {
    default: './.swag.config.js'
  })
  .option('-r, --dry-run', 'If dry run')
  .option('-f, --force', 'If force write')

cli.version(require('../../package.json')['version'])
cli.help()

const { options } = cli.parse()

const file = normalizePath(options['config'])
if (!existsSync(file)) {
  console.log(`config file ${options['config']} not exists.`)
  process.exit(0)
}
const config: rootConfig = require(file)

const run = async () => {
  const res = await gen(config)
  let data: string
  if (config.outType === 'yaml') {
    data = stringify(res)
  } else {
    data = JSON.stringify(res, null, 2)
  }

  if (options['dryRun']) {
    console.log(data)
    return
  }

  const outDir = normalizePath(config.outDir)
  if (!existsSync(outDir)) {
    mkdirp(outDir)
  }

  const outFile = join(outDir, `swagger.${config.outType}`)

  if (!options['force'] && existsSync(outFile)) {
    console.log(
      `${outFile} already exists, delete it or use -f option to replace it.`
    )
    return
  }

  writeFileSync(outFile, data)

  console.log('Done!')
}

run().catch(err => console.log(err.message))
