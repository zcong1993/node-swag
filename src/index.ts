import { parseDefinitions, parseFiles } from './parser'
import { RootConfig, Swag } from './types'
import { normalizeConfig } from './utils'

export * from './types'
export * from './parser'

export const gen = async (c: RootConfig): Promise<Swag> => {
  const config = normalizeConfig(c)
  // console.log(config)

  const res: Swag = {
    host: config.host,
    basePath: config.basePath,
    info: config.info,
    paths: {},
    swagger: '2.0'
  }
  if (config.model) {
    const definitions = await parseDefinitions(config.model)
    res.definitions = definitions
  }

  const paths = await parseFiles(config.files, config)

  res.paths = paths

  return res
}

// gen({
//   host: 'localhost:8080',
//   basePath: '/v1',
//   files: ['./zc/*.js'],
//   model: './zc/models.ts'
// })
//   .then(res => {
//     console.log(JSON.stringify(res))
//   })
