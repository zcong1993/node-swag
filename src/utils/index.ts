import { sync as fg } from 'fast-glob'
import { RootConfig } from '../types'

const validateConfig = (config: RootConfig) => {
  if (!config.host) {
    throw new Error('host is required')
  }
  if (!Array.isArray(config.files) || config.files.length === 0) {
    throw new Error('files should be array and not be none')
  }
}

export const normalizeConfig = (config: RootConfig): RootConfig => {
  validateConfig(config)

  if (!config.basePath) {
    config.basePath = '/'
  }

  if (!config.info) {
    config.info = {
      title: 'Swagger API',
      version: 'v0.0.0'
    }
  } else {
    config.info.title = config.info.title || 'Swagger API'
    config.info.version = config.info.version || 'v0.0.0'
  }

  if (config.outType !== 'yaml') {
    config.outType = 'json'
  }

  if (!config.outDir) {
    config.outDir = './.swag'
  }

  config.files.push('!**/node_modules')
  config.files = fg(config.files)

  return config
}
