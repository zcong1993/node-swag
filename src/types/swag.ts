import { Info, Security } from './rootConfig'
import { Definitions } from './definitions'

type Consumers =
  | 'application/json'
  | 'text/json'
  | 'application/xml'
  | 'text/xml'
  | 'application/x-www-form-urlencoded'
type Producers =
  | 'application/json'
  | 'text/json'
  | 'application/xml'
  | 'text/xml'
type ins = 'path' | 'query' | 'body'

interface Schema {
  $ref?: string
  type?: string
}

export interface SwaggerHttpEndpoint {
  tags: string[]
  summary?: string
  operationId?: string
  consumes: Consumers[]
  produces: Producers[]
  parameters: {
    name: string
    in: ins
    required: boolean
    description?: string
    type?: string
    maxLength?: number
    minLength?: number
  }[]
  respones: {
    [httpStatusCode: string]: {
      description: string
      schema?: Schema
    }
  }
  deprecated: boolean
}

export interface Swag extends Info {
  swagger: string
  host: string
  basePath: string
  definitions?: Definitions
  securityDefinitions?: {
    [key: string]: Security
  }
  paths: {
    [endpointPath: string]: {
      get?: SwaggerHttpEndpoint
      post?: SwaggerHttpEndpoint
      put?: SwaggerHttpEndpoint
      delete?: SwaggerHttpEndpoint
    }
  }
}
