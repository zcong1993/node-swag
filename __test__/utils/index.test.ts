import { normalizeConfig } from '../../src/utils'
import { RootConfig } from '../../src/types'

it('should works well', () => {
  const sample: RootConfig = {
    host: 'localhost:8080',
    basePath: '/v1',
    files: ['./src/*.js'],
    model: 'test.ts',
    outType: 'json'
  }

  expect(normalizeConfig(sample)).toMatchSnapshot()
})

it('should throw error', () => {
  const sample = {
    basePath: '/v1',
    files: ['./src/*.js'],
    model: 'test.ts',
    outType: 'json'
  }

  expect(() => normalizeConfig(sample as RootConfig)).toThrow()
})
