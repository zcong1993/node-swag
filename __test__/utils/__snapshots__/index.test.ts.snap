// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should works well 1`] = `
Object {
  "basePath": "/v1",
  "files": Array [],
  "host": "localhost:8080",
  "info": Object {
    "title": "Swagger API",
    "version": "v0.0.0",
  },
  "model": "test.ts",
  "outDir": "./.swag",
  "outType": "json",
}
`;
